File: ga/gpt3_5_result_2024_6/result/tree_gpt/26_tree.c
use std::ptr;

struct TreeNode {
    element: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

type SearchTree = Option<Box<TreeNode>>;
type Position = *mut TreeNode;

impl TreeNode {
    fn new(element: i32) -> Self {
        TreeNode {
            element,
            left: None,
            right: None,
        }
    }
}

fn make_empty(t: SearchTree) -> SearchTree {
    unsafe {
        if let mut t = t {
            make_empty(t.left.take());
            make_empty(t.right.take());
        }
    }
    None
}

fn find(x: i32, mut t: SearchTree) -> Option<Position> {
    while let Some(ptr) = t.as_deref_mut() {
        if x < ptr.element {
            t = ptr.left.take();
        } else if x > ptr.element {
            t = ptr.right.take();
        } else {
            return Some(ptr);
        }
    }
    None
}

fn find_min(mut t: SearchTree) -> Position {
    let mut current = t.as_mut().unwrap();
    while let Some(left) = current.left.as_deref_mut() {
        current = left;
    }
    let ptr: *mut TreeNode = current;
    ptr
}

fn find_max(mut t: SearchTree) -> Position {
    let mut current = t.as_mut().unwrap();
    while let Some(right) = current.right.as_deref_mut() {
        current = right;
    }
    let ptr: *mut TreeNode = current;
    ptr
}

fn insert(x: i32, t: SearchTree) -> SearchTree {
    match t {
        None => {
            Some(Box::new(TreeNode::new(x)))
        }
        Some(mut ptr) => {
            if x < ptr.element {
                ptr.left = insert(x, ptr.left.take());
            } else if x > ptr.element {
                ptr.right = insert(x, ptr.right.take());
            }
            Some(ptr)
        }
    }
}

fn delete(x: i32, t: SearchTree) -> SearchTree {
    match t {
        None => t,
        Some(mut node) if x < node.element => {
            node.left = delete(x, node.left.take());
            Some(node)
        }
        Some(mut node) if x > node.element => {
            node.right = delete(x, node.right.take());
            Some(node)
        }
        Some(node) if node.left.is_some() && node.right.is_some() => {
            let to_delete = find_min(node.right.take());
            node.element = unsafe { (*to_delete).element };
            node.right = delete(node.element, node.right.take());
            Some(node)
        }
        Some(node) => {
            let new_node = if let Some(ptr) = node.left.take() {
                ptr
            } else {
                node.right.take().unwrap_or_default()
            };
            Some(new_node)
        }
    }
}

fn retrieve(p: Position) -> i32 {
    let node = unsafe { &*p };
    node.element
}

fn main() {
    let mut t = make_empty(None);
    for i in 0..50 {
        let j = (7 * i) % 50;
        t = insert(j, t);
    }
    
    for i in 0..50 {
        if let Some(p) = find(i, t) {
            if retrieve(p) != i {
                println!("Error at {}", i);
            }
        }
    }

    for i in (0..50).step_by(2) {
        t = delete(i, t);
    }

    for i in (1..50).step_by(2) {
        if let Some(p) = find(i, t) {
            if retrieve(p) != i {
                println!("Error at {}", i);
            }
        }
    }

    for i in (0..50).step_by(2) {
        if let Some(_) = find(i, t) {
            println!("Error at {}", i);
        }
    }

    unsafe {
        let min = retrieve(find_min(t));
        let max = retrieve(find_max(t));
        println!("Min is {}, Max is {}", min, max);
    }
}

File: ga/glm4_result_2024_6/result/tree_glm/26_tree.rs
use std::ptr::null_mut;
use std::cmp::Ordering;

type SearchTree = *mut TreeNode;
type Position = *mut TreeNode;

struct TreeNode {
    element: i32,
    left: SearchTree,
    right: SearchTree,
}

fn make_empty(t: SearchTree) -> SearchTree {
    if t.is_null() {
        return null_mut();
    }
    make_empty(t.as_mut().unwrap().left);
    make_empty(t.as_mut().unwrap().right);
    unsafe {
        Box::from_raw(t);
    }
    null_mut()
}

fn find(x: i32, t: SearchTree) -> Position {
    if t.is_null() {
        return null_mut();
    }
    match x.cmp(&t.as_ref().unwrap().element) {
        Ordering::Less => find(x, t.as_mut().unwrap().left),
        Ordering::Greater => find(x, t.as_mut().unwrap().right),
        Ordering::Equal => t,
    }
}

fn find_min(t: SearchTree) -> Position {
    if t.is_null() {
        return null_mut();
    }
    if t.as_ref().unwrap().left.is_null() {
        t
    } else {
        find_min(t.as_mut().unwrap().left)
    }
}

fn find_max(t: SearchTree) -> Position {
    if t.is_null() {
        return null_mut();
    }
    let mut t = t;
    while t.as_ref().unwrap().right.is_some() {
        t = t.as_mut().unwrap().right;
    }
    t
}

fn insert(x: i32, t: SearchTree) -> SearchTree {
    if t.is_null() {
        let node = Box::new(TreeNode {
            element: x,
            left: null_mut(),
            right: null_mut(),
        });
        return Box::into_raw(node);
    }
    match x.cmp(&t.as_ref().unwrap().element) {
        Ordering::Less => {
            t.as_mut().unwrap().left = insert(x, t.as_mut().unwrap().left);
        }
        Ordering::Greater => {
            t.as_mut().unwrap().right = insert(x, t.as_mut().unwrap().right);
        }
        _ => {}
    }
    t
}

fn delete(x: i32, t: SearchTree) -> SearchTree {
    if t.is_null() {
        panic!("Element not found");
    }
    match x.cmp(&t.as_ref().unwrap().element) {
        Ordering::Less => {
            t.as_mut().unwrap().left = delete(x, t.as_mut().unwrap().left);
        }
        Ordering::Greater => {
            t.as_mut().unwrap().right = delete(x, t.as_mut().unwrap().right);
        }
        _ => {
            if t.as_ref().unwrap().left.is_some() && t.as_ref().unwrap().right.is_some() {
                let min_right = find_min(t.as_mut().unwrap().right);
                t.as_mut().unwrap().element = min_right.as_ref().unwrap().element;
                t.as_mut().unwrap().right = delete(t.as_mut().unwrap().element, t.as_mut().unwrap().right);
            } else {
                let mut tmp = t;
                if t.as_ref().unwrap().left.is_none() {
                    t = t.as_mut().unwrap().right;
                } else if t.as_ref().unwrap().right.is_none() {
                    t = t.as_mut().unwrap().left;
                }
                unsafe {
                    Box::from_raw(tmp);
                }
            }
        }
    }
    t
}

fn retrieve(p: Position) -> i32 {
    p.as_ref().unwrap().element
}

fn main() {
    let mut t: SearchTree = make_empty(null_mut());
    for i in 0..50 {
        t = insert(i, t);
    }
    for i in 0..50 {
        let p = find(i, t);
        if p.is_null() || retrieve(p) != i {
            println!("Error at {}", i);
        }
    }
    for i in 0..50 {
        if i % 2 == 0 {
            t = delete(i, t);
        }
    }
    for i in 1..50 {
        if i % 2 == 1 {
            let p = find(i, t);
            if p.is_null() || retrieve(p) != i {
                println!("Error at {}", i);
            }
        }
    }
    for i in 0..50 {
        if i % 2 == 0 {
            let p = find(i, t);
            if p.is_some() {
                println!("Error at {}", i);
            }
        }
    }
    println!("Min is {}, Max is {}", retrieve(find_min(t)), retrieve(find_max(t)));
}